recipe application final submission

RileyManda opened the pull request:
We have completed the task of building the Rails Recipe Application.

## Implementations 🌱 

## Project Setup:
> Created a new rails application
> Setup Linters and git actions/workflows


- ##  N+1 problem: Recipes

> Fixed N+1 problem: Recipes

<img width="669" alt="Screen Shot 2023-10-16 at 4 12 05 PM" src="https://github.com/RileyManda/recipe-app/assets/37291194/7b7a319d-03e5-41e4-aabe-6811e8da0e47">


> Fixed N+1 problem: Food

**Before**

<img width="365" alt="Screen Shot 2023-10-20 at 2 04 11 AM" src="https://github.com/RileyManda/recipe-app/assets/37291194/73af96f3-1d02-442f-8d9c-9dbbf0f316ec">


**Fix**
<img width="447" alt="Screen Shot 2023-10-20 at 2 09 37 AM" src="https://github.com/RileyManda/recipe-app/assets/37291194/a7954cb6-f8bf-4f5d-baac-f415af8473dd">


## Controllers and Model:

> Implemented user model
> Implemented food model
> Implemented recipe_food model
> Created database schema and migrations using rails orm
> Implemented Food controller and actions
> Implemented Recipe controller and actions
> Implemented RecipeFood controller and actions


## Views
> ingredients table
> recipes view
> recipes view
> Add recipe and food forms
> Display Ingredients Table
> display foods list
> add food 

## Validations
> Implemented validations using cancan on foods and recipes
> Remove recipe validations:If user is recipe owner-can remove recipe
> Remove food validations:If user is food owner-can view food and can delete food

## Tests

### Unit Tests
> Added Rspec rails gem
> Wrote unit tests for recipes
> Wrote unit tests for foods 

<img width="459" alt="Screen Shot 2023-10-20 at 2 14 29 AM" src="https://github.com/RileyManda/recipe-app/assets/37291194/f34edcf9-2b53-41c1-b8b7-b29e7d775a8b">
<img width="566" alt="Screen Shot 2023-10-20 at 2 12 39 AM" src="https://github.com/RileyManda/recipe-app/assets/37291194/86250ef8-571f-4ec0-a962-95cafa1a3bef">
<img width="564" alt="Screen Shot 2023-10-20 at 2 13 36 AM" src="https://github.com/RileyManda/recipe-app/assets/37291194/9d892a31-8105-4e0c-b077-b565c226ffbb">





### Integration tests
> Added capybara gem
> Wrote integration tests for recipe and recipe foods using Capybara

 In addition the following general requirements have been fulfilled:
 
### General requirements
> Used gitflow for this project
> Conducted code reviews on contributions
> Made sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
> Made sure that I have used correct flow [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
> Made sure that I have documented my work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
> Followed list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).


Please kindly review this implementation.

Thank you 🙏🏻

tresorsawasawa commented on app/views/recipe/index.html.erb:
```diff
@@ -0,0 +1,28 @@
+<div class="container">
+  <div class="row">
+    <div class="col-md-12">
+      <h2 class="mb-3 text-center">Recipes</h2>
+      <%= button_to 'Add Recipe', new_recipe_path, method: :get, class: 'btn btn-primary btn-sm' %>
+      <ul class="list-group">
+        <% if @recipes.any? %>
+        <% @recipes.each do |recipe| %>
+          <li class="list-group-item d-flex justify-content-between align-items-center">
+            <div>
+             <h3><%= link_to recipe.name, recipe_path(recipe) %></h3>
+            <% if can? :manage, recipe %>
+  <%= button_to "Remove", recipe, method: :delete, data: { confirm: "Are you sure?" }, class: "btn btn-danger btn-sm" %>
+<% end %>
+            </div>
+            <p class="text-muted"><%= recipe.description %></p>
+
+          </li>
```
- [ **OPTIONAL** ] Good job as your design follow the **[provided wireframe](https://raw.githubusercontent.com/microverseinc/curriculum-rails/main/recipe-app/images/recipe_index.png?token=GHSAT0AAAAAAB56UIU3ZZPWCT3YG3NBDG52ZJRMAJQ)**. However, I think it would be better if you improve again your design to look more professional as now when the text description is long text are overlapping the recipe box and text-alignment are set to right by default, such as the image below:
![image](https://github.com/RileyManda/recipe-app/assets/83042742/a8412a96-1b1a-40f3-af39-9fd99f671d64)

tresorsawasawa commented on spec/requests/shopping_spec.rb:
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+RSpec.describe 'Shoppings', type: :request do
+  describe 'GET /index' do
+    it 'returns http success' do
+      get '/shopping/index'
+      expect(response).to have_http_status(:success)
+    end
+  end
```
- [ ] As required for the project, kindly ensure that all your tests are passing as there some of them that are failing.
![image](https://github.com/RileyManda/recipe-app/assets/83042742/cf72c0cf-b737-422d-a495-218688136429)

tresorsawasawa commented on app/views/recipe/show.html.erb:
```diff
@@ -0,0 +1,57 @@
+  </div>
+  <p>Cooking Time: <%= @recipe.cooking_time %> minutes</p>
+
+  <div class="col-md-6">
+     <%= link_to "Generate Shopping list", new_recipe_food_path(recipe_id: @recipe.id), class: 'btn btn-primary btn-sm' %>
+    <%= link_to "Add Ingredient", new_recipe_food_path(recipe_id: @recipe.id), class: 'btn btn-primary btn-sm' %>
```
- [ ] As required for the project **Generate Shopping list** to show the list of food that is missing for all recipes of the logged-in user and count the total food items and total price of the missing food. Kindly ensure that when clicking on the **Generate Shopping list** button it executes its expected task. Right now, the all above buttons are doing same thing of only adding **Gradients**. Please fix that.

tresorsawasawa reviewed (changes requested):
# STATUS: REQUIRED CHANGES :recycle::recycle: 

Hi @RileyManda and @yin-ka 
Good job so far! I know you worked hard to implement this project, I really appreciate the effort you put in! But there are some issues that you still need to work on to go to the next project but you are almost there! 

![almost-there](https://user-images.githubusercontent.com/57562869/143511358-9cde7daf-4ad0-4db2-8f15-18b54288ff91.gif)

## To highlight :champagne: 
- All linters are passing :heavy_check_mark: 
- Follow GitHub flow :heavy_check_mark: 
- You have fixed  fix N+1 queries problem :heavy_check_mark: 
- Good associations :heavy_check_mark:
- You Wrote unit and integrations tests using Rspec :heavy_check_mark: 
- Used validations for models :heavy_check_mark: 
- You added authentication with Devise :heavy_check_mark:  

## Required changes :recycle: :recycle: 
- [ ] Check the comments under the review.

## Optional suggestions

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

_Feel free to leave any questions or comments in the PR thread if something is not 100% clear. 
Please ping me [@tresorsawasawa](https://github.com/tresorsawasawa) when you comment so I can receive the notification or use slack with the same name._

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

RileyManda commented:
We have completed the task of making the requested changes to the project:

## Iterations:

- Updated recipe card to avoid text overflow

- Fix failing tests

- Updated readme with licence

- Implemented shopping list logic

RileyManda commented:
We have completed the task of making the requested changes to the project:

## Iterations:

- Updated recipe card to avoid text overflow

- Fix failing tests

- Updated readme with licence

- Implemented functional shopping list logic:




In addition  following general requirements have been fulfilled:
 
### General requirements
> Used gitflow for this project
> Conducted code reviews on contributions
> Made sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
> Made sure that I have used correct flow [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
> Made sure that I have documented my work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
> Followed list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).


Please kindly review this implementation.

Thank you 🙏🏻

fobadara commented on test/models/food_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class FoodTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

RileyManda replied:
Thank you for the suggested changes @fobadara 
We will make the changes and submit for review.Thank you

fobadara commented on test/models/recipe_food_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class RecipeFoodTest < ActiveSupport::TestCase
+  # test "the truth" do
+  #   assert true
+  # end
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on test/controllers/users_controller_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on test/controllers/recipes_controller_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class RecipeControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on test/controllers/recipe_foods_controller_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class RecipeFoodControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on test/channels/application_cable/connection_test.rb:
```diff
@@ -0,0 +1,11 @@
+require 'test_helper'
+
+class ApplicationCable::ConnectionTest < ActionCable::Connection::TestCase
+  # test "connects with cookies" do
+  #   cookies.signed[:user_id] = 42
+  #
+  #   connect
+  #
+  #   assert_equal connection.user_id, "42"
+  # end
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on test/controllers/foods_controller_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
+
+class FoodControllerTest < ActionDispatch::IntegrationTest
+  # test "the truth" do
+  #   assert true
+  # end
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on spec/helpers/shopping_helper_spec.rb:
```diff
@@ -0,0 +1,15 @@
+# Specs in this file have access to a helper object that includes
+# the ShoppingHelper. For example:
+#
+# describe ShoppingHelper do
+#   describe "string concat" do
+#     it "concats two strings with spaces" do
+#       expect(helper.concat_strings("this","that")).to eq("this that")
+#     end
+#   end
+# end
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on db/migrate/20231010153137_add_devise_to_users.rb:
```diff
@@ -0,0 +1,51 @@
+
+    add_index :users, :email,                unique: true
+    add_index :users, :reset_password_token, unique: true
+    # add_index :users, :confirmation_token,   unique: true
+    # add_index :users, :unlock_token,         unique: true
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on db/migrate/20231010153137_add_devise_to_users.rb:
```diff
@@ -0,0 +1,51 @@
+      t.datetime :confirmation_sent_at
+      t.string   :unconfirmed_email # Only if using reconfirmable
+
+      ## Lockable
+      # t.integer  :failed_attempts, default: 0, null: false # Only if lock strategy is :failed_attempts
+      # t.string   :unlock_token # Only if unlock strategy is :email or :both
+      # t.datetime :locked_at
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on db/migrate/20231010153137_add_devise_to_users.rb:
```diff
@@ -0,0 +1,51 @@
+      ## Rememberable
+      t.datetime :remember_created_at
+
+      ## Trackable
+      # t.integer  :sign_in_count, default: 0, null: false
+      # t.datetime :current_sign_in_at
+      # t.datetime :last_sign_in_at
+      # t.string   :current_sign_in_ip
+      # t.string   :last_sign_in_ip
```
[REQUIRED] As suggested by the previous reviewer kindly remove all commented-out unused code.

fobadara commented on spec/models/shopping_spec.rb:
```diff
@@ -1,5 +1,5 @@
-end
+# RSpec.describe Shopping, type: :model do
+#   pending "add some examples to (or delete) #{__FILE__}"
+# end
```
[REQUIRED] Kindly remove all commented-out code to make your code more professional

fobadara reviewed (changes requested):
Hi team, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which haven't been addressed yet.

![REDO](https://user-images.githubusercontent.com/62938735/260208741-25cd532c-dea0-4827-bb4f-5412e8493916.gif)

## Highlights
✅ Descriptive README
✅ Descriptive PR summary
✅ Clean code

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you consider them as they can improve your code. Some of them were simply missed by the previous reviewer and addressing them will improve your application._


Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

RileyManda commented:
We have completed the task of making the requested changes to the project:

## Iterations:

- Removed unnecessary rails boilerplate and commented out code from the project





In addition  following general requirements have been fulfilled:
 
### General requirements
> Used gitflow for this project
> Conducted code reviews on contributions
> Made sure that there are [no linter errors](https://github.com/microverseinc/linters-config).
> Made sure that I have used correct flow [Gitflow](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/git-github/articles/gitflow.md).
> Made sure that I have documented my work [in a professional way](https://github.com/microverseinc/curriculum-transversal-skills/blob/main/documentation/articles/professional_repo_rules.md).
> Followed list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).


Please kindly review this implementation.

Thank you 🙏🏻

mckent05 reviewed (approved):
Hi @RileyManda, 

# STATUS: APPROVED ✔️ ✔️ 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations and well done! 🎉 
![approved](https://user-images.githubusercontent.com/73607512/207264357-1b05ce6a-95f2-4d31-9c26-888c6ddfa124.gif)


## Highlights
- All linter checks are passing ✔️ ✔️ 
- Your PR is professional 👏 👏 
- Good job completing your capstone project 👍 👍 
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
